### PR TITLE
This is the starting support for directives as Nadel instructions.

### PR DESCRIPTION
**Pull request by bbakerman**
This breaks Nadel so that the outer caller has to give it syntax by named "service"

That is `service X { ... }` no longer is the source of service name.

Instead you must give it a Map<String,Reader> of DSL.

BUT the old syntax still works.  This preserves backwards compatibility

Most of the tests are now like this

`def ndsl= [ ServiceName : " service ServiceName { ... } " ]

So some service could be

MyService.nadel

```
   type Query {
      someField : String @RenamedFrom(value : "oldFieldName")
   }
``

### CODE REVIEW

File: src/main/java/graphql/nadel/Nadel.java [hunk 1]
```diff
@@ -329,13 +346,30 @@ public static Builder newNadel() {
...
+            requireNonNull(nsdl);
+            this.serviceNDSLs.put(serviceName, nsdl);
             return this;
         }
 
-        public Builder dsl(String nsdl) {
-            return dsl(new StringReader(requireNonNull(nsdl)));
+        public Builder dsl(String serviceName, String nsdl) {
+            return dsl(serviceName, new StringReader(requireNonNull(nsdl)));
+        }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/Nadel.java:357]: These are only useful for testing

File: src/main/java/graphql/nadel/NadelAntlrToLanguage.java [hunk 2]
```diff
@@ -72,13 +72,19 @@ protected void addCommonData(NodeBuilder nodeBuilder, ParserRuleContext parserRu
...
         StitchingDsl.Builder builder = StitchingDsl.newStitchingDSL();
         addCommonData(builder, ctx);
-        List<ServiceDefinition> serviceDefinitions = ctx.serviceDefinition().stream()
-                .map(this::createServiceDefinition)
-                .collect(Collectors.toList());
-        builder.serviceDefinitions(serviceDefinitions);
+        if (ctx.serviceDefinition() != null) {
+            ServiceDefinition serviceDefinition = createServiceDefinition(ctx.serviceDefinition());
+            builder.serviceDefinition(serviceDefinition);
+        }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/NadelAntlrToLanguage.java:78]: There can be ONLY 1 service inside a parsed text stream!

File: src/test/groovy/graphql/nadel/NSDLParserTest.groovy [hunk 3]
```diff
@@ -48,30 +49,6 @@ class NSDLParserTest extends Specification {
         assertNodesHaveIds(stitchingDSL.children)
     }
 
-    def "two services"() {
```
Comment (bbakerman) [inline src/test/groovy/graphql/nadel/NSDLParserTest.groovy:51]: No longer valid

File: src/test/groovy/graphql/nadel/NSDLParserTest.groovy [hunk 4]
```diff
@@ -338,6 +315,31 @@ class NSDLParserTest extends Specification {
         then:
         astAsMap(stitchingDSL).size() > 0
     }
+
+    def "parser allows normal .graphqls syntax"() {
```
Comment (andimarek) [inline src/test/groovy/graphql/nadel/NSDLParserTest.groovy:319]: you mean SDL :-)

**Review by andimarek — APPROVED** [review]

File: src/main/java/graphql/nadel/util/Util.java [hunk 5]
```diff
@@ -40,25 +44,45 @@
...
-        for (Definition definition : serviceDefinition.getTypeDefinitions()) {
-            definitionRegistry.add((SDLDefinition) definition);
+        for (Definition<?> definition : serviceDefinition.getTypeDefinitions()) {
+            definitionRegistry.add((SDLDefinition<?>) definition);
         }
         return definitionRegistry;
     }
 
+
+    public static ServiceDefinition mkServiceDefinition(String serviceName, StitchingDsl stitchingDsl) {
```
[the remark this answers is not in the document]
Comment (bbakerman) [reply src/main/java/graphql/nadel/util/Util.java]: fixed

File: src/test/groovy/graphql/nadel/NSDLParserTest.groovy (hunk 4 above)
Comment (bbakerman) [reply src/test/groovy/graphql/nadel/NSDLParserTest.groovy]: fixed

File: src/main/java/graphql/nadel/schema/NadelDirectives.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+import static graphql.language.TypeName.newTypeName;
+import static graphql.nadel.dsl.RemoteArgumentSource.SourceType.CONTEXT;
+import static graphql.nadel.dsl.RemoteArgumentSource.SourceType.FIELD_ARGUMENT;
+import static graphql.nadel.dsl.RemoteArgumentSource.SourceType.OBJECT_FIELD;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+
+public class NadelDirectives {
+
+    static final DirectiveDefinition HIDE_DIRECTIVE_DEFINITION;
```
Comment (andimarek) [inline src/main/java/graphql/nadel/schema/NadelDirectives.java:49]: we should think about this naming for a second and make sure we consider directives for both use case:
directives for the overall schema (when you still have an underlying one) and the directives if you only have one underlying schema.

Comment (andimarek) [reply src/main/java/graphql/nadel/schema/NadelDirectives.java:49]: can we remove the hide directive for now and revisit it again when we need it?

File: src/main/java/graphql/nadel/schema/NadelDirectives.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+                        newInputValueDefinition()
+                                .name("from")
+                                .description(createDescription("The type to be renamed"))
+                                .type(nonNull("String"))
+                                .build())
+                .build();
+
+        HIDE_DIRECTIVE_DEFINITION = DirectiveDefinition.newDirectiveDefinition()
+                .name("hide")
+                .directiveLocation(newDirectiveLocation().name(FIELD_DEFINITION.name()).build())
```
Comment (andimarek) [inline src/main/java/graphql/nadel/schema/NadelDirectives.java:75]: we should make this simpler in GraphQL Java :-)

**Review by andimarek — CHANGES_REQUESTED** [review]
see my comment regarding hide directive

File: src/main/java/graphql/nadel/schema/NadelDirectives.java (hunk 6 above)
Comment (bbakerman) [reply src/main/java/graphql/nadel/schema/NadelDirectives.java]: removed

**Pull request merged**